### PR TITLE
LisyRuiz - Q's - Media Ranker

### PR DESCRIPTION
**Pull request by LeesyRuiz**
# Media Ranker

so..... this is basically a disaster. 
1) i don't have any tests so I have assigned myself that responsibility for betsy.
2) most recent heroku push failed so i just rolledback to v7 for the sake of this pull request. 
4) i had my voting logic working in the controller but haven't established the links yet on my working site. 
5) by monday i will have an more complete site completed & can then better answer the questions below. 

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a custom model method you wrote. |   |
| Describe how you approached testing that model method. What edge cases did you come up with? |  |
| Describe an edge case test you wrote for a controller |  |
| What are `session` and `flash`? What is the difference between them? |  |
| Describe a controller filter you wrote. |  |

| What was one thing that you gained more clarity on through this assignment? | how much more intricate rails is than i thought.  |

| What is the Heroku URL of your deployed application | https://lisymediaranker.herokuapp.com/  |
| Do you have any recommendations on how we could improve this project for the next cohort? |  |

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
+class UsersController < ApplicationController
+  before_action :set_user, only: [:show, :edit, :update, :destroy]
+
+
+  def index
+    @users = User.all
+    @username = User.find(session[:user_id]).name
```
Comment (droberts-sea) [inline app/controllers/users_controller.rb:7]: Why do you find the `@username` in `index`? I don't think there's anything specific that needs it in the users index, and if it's something you need for the whole site you should put it in `ApplicationController`.

File: app/controllers/votes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  # before_action :authenticate_user!
+  before_action :set_work
+
+  def create
+    @work.votes.where(user_id: current_user.id).first_or_create
+
+    respond_to do |format|
+      format.html { redirect_to @work }
+      format.js
```
Comment (droberts-sea) [inline app/controllers/votes_controller.rb:11]: What's going on here? I'm guessing this is something you took from StackOverflow - we won't see `respond_to` until we start building our own APIs in a couple weeks, and this doesn't really solve a problem I would expect you to run into on this project. I appreciate the initiative, but you need to make sure you fully understand *why* you're adding what you add.

File: app/views/users/_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+<%= form_for(user) do |f| %>
+  <% if user.errors.any? %>
+    <div id="error_explanation">
```
Comment (droberts-sea) [inline app/views/users/_form.html.erb:3]: Good work presenting validation errors to users!

File: config/routes.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  root "welcome#index"
+
+
+  get 'welcome/menu'
+
+  get 'welcome/login'
+  post 'welcome/login'
+
+  get 'welcome/logout'
+  delete 'welcome/logout'
```
Comment (droberts-sea) [inline config/routes.rb:14]: When writing routes, you should always include the `to: 'controller#action'` argument - this makes it very clear what controller action the request will be routed to. It's also usually wise to explicitly name the URI using the `as: 'name'` argument.

File: config/routes.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+  get 'logout' => 'welcome#destroy'
+
+
+
+  resources :works do
+    member do
+       resource :vote
```
Comment (droberts-sea) [inline config/routes.rb:22]: I don't know that the full set of resources makes sense for `vote` - you should be able to get away with just one route for upvoting.

File: config/routes.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+      get :delete
+      get :vote
+      post :vote
+
+    end
+  end
+
+
+  resources :users
```
Comment (droberts-sea) [inline config/routes.rb:31]: Do you need the full set of resources for users?

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  def lookup_user
+    unless session[:user_id].nil?
+      @user =User.find_by(id: session[:user_id])
+    end
+  end
+
+
+  def require_login
+    lookup_user
+    if @current_user.nil?
```
Comment (droberts-sea) [inline app/controllers/application_controller.rb:26]: In `lookup_user` you set `@user`, but in `require_login` you look for `@current_user`, so `require_login` will always fail.

File: app/controllers/works_controller.rb [hunk 8]
```diff
@@ -0,0 +1,89 @@
+class WorksController < ApplicationController
+
+  before_action :require_login, except [:index]
```
Comment (droberts-sea) [inline app/controllers/works_controller.rb:3]: The project spec says you should only `require_login` for voting, not for looking at media.

**Comment by droberts-ada** [comment]
# Media Ranker

Before I start I'd like to acknowledge a couple things:
- MediaRanker is a big, complex project where we don't expect everybody to complete everything
- The main work weekend for this project was right in the middle of your personal crisis

What you've submitted here is a little concerning - much of the site doesn't work out of the box, and I'm worried that some of the learning goals were missed (particularly complex model logic, model and controller testing, and imitating an existing visual design). However, given the circumstances I think we can let it slide a little, so long as you take this feedback as an opportunity to improve your understanding of the material.

## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
| **Core Requirements** |   |
| Git hygiene | yes |
| Comprehension questions	| yes |
| **General** |  |
| Rails fundamentals (RESTful routing, use of named paths) | Your routes could use some attention - you've created many that you're not using. In general, don't add a route to your application until you're ready to implement it. |
| Semantic HTML | no |
| Errors are reported to the user | yes |
| Business logic lives in the models | Some - the `votes?` method is a good idea here. Other good candidates would be the logic to get the top 10 media for a given category sorted by vote count, and the logic to find the spotlight media. |
| Model testing | no |
| Controller testing | good start - I see success cases for many of your controller actions, but not a lot of failure cases. |
| **Wave 1 - Media** |  |
| Splash page shows the three media categories | yes |
| Basic CRUD operations on media are present and functional | I was not able to get this to work, even with some simple edits to your code - seems like by adding user login you have broken everything else. |
| **Wave 2 - Users and Votes** |  |
| Users can log in and log out | Good start. Looks like your login and logout functions are fine, but there's a disconnect between what these do and what the rest of your app is looking for. |
| The ID of the current user is stored in the session | yes |
| Individual user pages and the user list are present | yes |
| A user cannot vote for the same media more than once | Could not get voting to work |
| All media lists are ordered by vote count | Could not get voting to work |
| Splash page contains a media spotlight | no |
| Media pages contain lists of voting users | no |
| **Wave 3 - Styling** |  |
| Foundation is used appropriately | good start |
| Look and feel is similar to the original | no - all the extra stuff on the homepage (animated .gifs, etc) is very creative, but it's also not what we asked for |
|  **Overall** |  |

Particular things I noticed:
- Your indentation is sloppy! Having good habits around indentation makes your code *much* easier to read, which in turn makes it easier to debug. There's a reason we set up a keyboard shortcut for this at the beginning of the cohort.
- Your `.env` file is included on github. For this project there wasn't anything in it so it's OK, but in general you should try and build a habit of including this in your `.gitignore` file.
- Much of the site, including pages you shouldn't need to be logged in to see, uses the `user_id` from the session, so it breaks when the user hasn't logged in yet. This made your site very difficult to test.

For bEtsy and in your 1-on-1 work with Mariano, please spend some time focusing on model logic: writing and testing validations, custom instance methods and class methods for interesting database selection.

Let me know if you have any questions or need some more direction on this, or if you'd just like to meet 1-on-1.
